rules/opentitan.bzl

```diff
@@ -0,0 +1,203 @@
+        ctx.actions.run_shell(
+            outputs = [disassembly],
+            inputs = [src] + cc_toolchain.all_files.to_list(),
+            arguments = [
+                cc_toolchain.objdump_executable,
+                src.path,
+                disassembly.path,
+            ],
+            command = "$1 --disassemble --headers --line-numbers --source $2 > $3",
+        )
+        return [DefaultInfo(files = depset(outputs), data_runfiles = ctx.runfiles(files = outputs))]
+
+elf_to_disassembly = rule(
+    implementation = _elf_to_disassembly,
+    cfg = _platforms_transition,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "platform": attr.string(default = OPENTITAN_PLATFORM),
+        "_cc_toolchain": attr.label(default = Label("@bazel_tools//tools/cpp:current_cc_toolchain")),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+    },
+    toolchains = ["@rules_cc//cc:toolchain_type"],
+    incompatible_use_toolchain_transition = True,
+)
+
+def _elf_to_scrambled(ctx):
+    outputs = []
+    for src in ctx.files.srcs:
+        scrambled = ctx.actions.declare_file("{}.scr.40.vmem".format(src.basename))
+        outputs.append(scrambled)
+        ctx.actions.run(
+            outputs = [scrambled],
+            inputs = [
+                src,
+                ctx.files._tool[0],
+                ctx.files._config[0],
+            ],
+            arguments = [
+                ctx.files._config[0].path,
+                src.path,
+                scrambled.path,
+            ],
+            executable = ctx.files._tool[0].path,
+        )
+        return [DefaultInfo(files = depset(outputs), data_runfiles = ctx.runfiles(files = outputs))]
+
+elf_to_scrambled = rule(
+    implementation = _elf_to_scrambled,
+    cfg = _platforms_transition,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "platform": attr.string(default = OPENTITAN_PLATFORM),
+        "_tool": attr.label(default = "//hw/ip/rom_ctrl/util:scramble_image.py", allow_files = True),
+        "_config": attr.label(default = "//hw/top_earlgrey:data/autogen/top_earlgrey.gen.hjson", allow_files = True),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+    },
+)
+
+def opentitan_binary(
+        name,
+        platform = OPENTITAN_PLATFORM,
+        per_device_deps = {
+            "verilator": ["//sw/device/lib/arch:sim_verilator"],
+            "dv": ["//sw/device/lib/arch:sim_dv"],
+            "fpga_nexysvideo": ["//sw/device/lib/arch:fpga_nexysvideo"],
+            "cw310": ["//sw/device/lib/arch:fpga_cw310"],
+        },
+        output_bin = True,
+        output_disassembly = True,
+        output_scrambled = False,
+        **kwargs):
+    """A helper macro for generating OpenTitan binary artifacts.
```

Comment:
Can you include a description of the rules generated?

Reply:
@mcy Sorry, I don't understand what this is asking.

Reply:
This will emit a whole bunch of rules like `name + "_elf"`. Documenting those is important; I've previously dealt with macros that did this kind of thing but conveniently didn't document it...

Reply:
This macro is going to emit the following:

For each key in `per_device_deps`, it will emit:
- A cc_binary rule named `$name_$key`.
- An ELF file specifically of the format `elf32-little` named `$name_$key_elf`.
- Then, depending on the various output flags, it may emit:
  - A binary transformation named `$name_$key_$bin`.
  - A text file which is the disassembly named `$name_$key_dis`.
  - A scrambled binary (for ROM) named `$name_$key_scr`.

Lastly, this rule will emit a filegroup named `$name` with all of the previously generated rules as its `srcs`.

Reply:
I've added a summary to the comment here.